File: doordeck-sdk/src/mingwMain/resources/csharp/Wrapper/Account.cs
using Doordeck.Headless.Sdk.Model;
using Doordeck.Headless.Sdk.Model.Responses;

namespace Doordeck.Headless.Sdk.Wrapper;

    
public class Account
{
    public Task<TokenResponse> RefreshToken(string? refreshToken = null) =>
        Dispatcher.Call<TokenResponse>("account.refreshToken", new { refreshToken });

    public Task<object> Logout() =>
        Dispatcher.Call<object>("account.logout"); 

    public Task<RegisterEphemeralKeyResponse> RegisterEphemeralKey(byte[]? publicKey = null, byte[]? privateKey = null) =>
        Dispatcher.Call<RegisterEphemeralKeyResponse>("account.registerEphemeralKey", new { publicKey, privateKey });

    public Task<RegisterEphemeralKeyWithSecondaryAuthenticationResponse> RegisterEphemeralKeyWithSecondaryAuthentication(byte[]? publicKey = null, TwoFactorMethod? method = null) =>
        Dispatcher.Call<RegisterEphemeralKeyWithSecondaryAuthenticationResponse>("account.registerEphemeralKeyWithSecondaryAuthentication", new { publicKey, method });

    public Task<RegisterEphemeralKeyResponse> VerifyEphemeralKeyRegistration(string code, byte[]? publicKey = null, byte[]? privateKey = null) =>
        Dispatcher.Call<RegisterEphemeralKeyResponse>("account.verifyEphemeralKeyRegistration", new { code, publicKey, privateKey });

    public Task<object> ReverifyEmail() =>
        Dispatcher.Call<object>("account.reverifyEmail");

    public Task<object> ChangePassword(string oldPassword, string newPassword) =>
        Dispatcher.Call<object>("account.changePassword", new { oldPassword, newPassword });

    public Task<UserDetailsResponse> GetUserDetails() =>
        Dispatcher.Call<UserDetailsResponse>("account.getUserDetails");

    public Task<object> UpdateUserDetails(string displayName) =>
        Dispatcher.Call<object>("account.updateUserDetails", new { displayName });

    public Task<object> DeleteAccount() =>
        Dispatcher.Call<object>("account.deleteAccount");
}

File: doordeck-sdk/src/appleMain/kotlin/com/doordeck/multiplatform/sdk/model/data/Account.kt
package com.doordeck.multiplatform.sdk.model.data

import com.doordeck.multiplatform.sdk.model.common.TwoFactorMethod
import com.doordeck.multiplatform.sdk.model.responses.RegisterEphemeralKeyResponse
import com.doordeck.multiplatform.sdk.model.responses.RegisterEphemeralKeyWithSecondaryAuthenticationResponse
import com.doordeck.multiplatform.sdk.model.responses.TokenResponse
import com.doordeck.multiplatform.sdk.model.responses.UserDetailsResponse

data class Token(
    val authToken: String,
    val refreshToken: String
)

data class UserDetails(
    val email: String,
    val displayName: String? = null,
    val emailVerified: Boolean,
    val publicKey: String
)

data class RegisterEphemeralKey(
    val certificateChain: List<String>,
    val userId: String
)

data class RegisterEphemeralKeyWithSecondaryAuthentication(
    val method: TwoFactorMethod
)

internal fun TokenResponse.toToken(): Token = Token(
    authToken = authToken,
    refreshToken = refreshToken
)

internal fun UserDetailsResponse.toUserDetails(): UserDetails = UserDetails(
    email = email,
    displayName = displayName,
    emailVerified = emailVerified,
    publicKey = publicKey
)

internal fun RegisterEphemeralKeyResponse.toRegisterEphemeralKey(): RegisterEphemeralKey = RegisterEphemeralKey(
    certificateChain = certificateChain,
    userId = userId
)

internal fun RegisterEphemeralKeyWithSecondaryAuthenticationResponse.toRegisterEphemeralKeyWithSecondaryAuthentication(): RegisterEphemeralKeyWithSecondaryAuthentication = RegisterEphemeralKeyWithSecondaryAuthentication(
    method = method
)